File: Serenity.Script.Core/SlickGrid/Formatters/FileDownloadFormatter.cs
using Serenity.ComponentModel;
using System.Collections.Generic;

namespace Serenity
{
    public class FileDownloadFormatter : ISlickFormatter, IInitializeColumn
    {
        public string Format(SlickFormatterContext ctx)
        {
            var dbFile = ctx.Value as string;

            if (string.IsNullOrEmpty(dbFile))
                return "";

            var downloadUrl = DbFileUrl(dbFile);
            var originalName = !string.IsNullOrEmpty(OriginalNameProperty) ?
                ctx.Item[OriginalNameProperty] as string : null;
            originalName = originalName ?? "";

            var text = string.Format(DisplayFormat ?? "{0}",
                originalName,
                dbFile,
                downloadUrl);

            return "<a class='file-download-link' target='_blank' " + 
                "href='" + Q.HtmlEncode(downloadUrl) + "'>" +
                Q.HtmlEncode(text) +
                "</a>";
        }

        public static string DbFileUrl(string filename)
        {
            filename = (filename ?? "").Replace("\\", "/");
            return Q.ResolveUrl("~/upload/") + filename;
        }

        [Option]
        public string DisplayFormat { get; set; }

        [Option]
        public string OriginalNameProperty { get; set; }

        public void InitializeColumn(SlickColumn column)
        {
            column.ReferencedFields = column.ReferencedFields ?? new List<string>();

            if (!string.IsNullOrEmpty(OriginalNameProperty))
            {
                column.ReferencedFields.Add(OriginalNameProperty);
                return;
            }
        }
    }
}

File: packages/corelib/src/ui/formatters/filedownloadformatter.ts
import { Column, FormatterContext } from "@serenity-is/sleekgrid";
import { faIcon, formatterTypeInfo, htmlEncode, iconClassName, registerType, resolveUrl, stringFormat } from "../../base";
import { replaceAll } from "../../compat";
import { Formatter } from "../../slick";
import { IInitializeColumn } from "./iinitializecolumn";

export class FileDownloadFormatter implements Formatter, IInitializeColumn {
    static typeInfo = formatterTypeInfo("Serenity.IFileDownloadFormatter", [IInitializeColumn]); static { registerType(this); }

    constructor(public readonly props: { displayFormat?: string, originalNameProperty?: string, iconClass?: string } = {}) {
        this.props ??= {};
    }

    format(ctx: FormatterContext): string {
        var dbFile = ctx.value as string;
        if (!dbFile)
            return '';

        var downloadUrl = FileDownloadFormatter.dbFileUrl(dbFile);
        var originalName = this.originalNameProperty ?
            ctx.item[this.originalNameProperty] as string : null;

        originalName = (originalName ?? '');
        var text = stringFormat((this.displayFormat ?? '{0}'),
            originalName, dbFile, downloadUrl);

        var iconClass = iconClassName(this.iconClass ?? faIcon("download"));

        return "<a class='file-download-link' target='_blank' href='" +
            htmlEncode(downloadUrl) + "'><i class='" + iconClass + "'></i> " + htmlEncode(text) + '</a>';
    }

    static dbFileUrl(filename: string): string {
        filename = replaceAll((filename ?? ''), '\\', '/');
        return resolveUrl('~/upload/') + filename;
    }

    initializeColumn(column: Column): void {
        column.referencedFields = column.referencedFields || [];
        if (this.originalNameProperty) {
            column.referencedFields.push(this.originalNameProperty);
            return;
        }
    }

    get displayFormat() { return this.props.displayFormat; }
    set displayFormat(value) { this.props.displayFormat = value; }

    get originalNameProperty() { return this.props.originalNameProperty; }
    set originalNameProperty(value) { this.props.originalNameProperty = value; }

    get iconClass() { return this.props.iconClass; }
    set iconClass(value) { this.props.iconClass = value; }
}